fulcio.proto

```diff
@@ -49,16 +49,21 @@ message CreateSigningCertificateRequest {
      * Identity information about who possesses the private / public key pair presented
      */
     Credentials credentials    = 1 [(google.api.field_behavior) = REQUIRED];
-    /*
-     * The public key to be stored in the requested certificate
-     */
-    PublicKey public_key       = 2 [(google.api.field_behavior) = REQUIRED];
-    /*
-     * Proof that the client possesses the private key; must be verifiable by provided public key
-     *
-     * This is a currently a signature over the `sub` claim from the OIDC identity token
-     */
-    bytes proof_of_possession  = 3 [(google.api.field_behavior) = REQUIRED];
+    oneof key {
+        /*
+         * The public key to be stored in the requested certificate along with a signed
+         * challenge as proof of possession of the private key.
+         */
+        PublicKeyRequest public_key_request = 2 [(google.api.field_behavior) = REQUIRED];
```

Comment:
What is the format of this key?

Will we ever want to accept multiple formats of public key (we probably do)? Right now the implementation just tries to decode in a couple of different ways IIRC, but if we add more formats we probably want to specify that in the API. Should we separate out those code paths now? I think I'd rather have users say "this is PEM" or "this is DER" than just trying to parse that out ourselves.

I know this isn't an issue at all right now, but I could imagine it becoming one soon.

Reply:
Yep, currently we support PEM and DER encoded keys. My concern is a client might not be aware of the differences between PEM or DER encodings. Even that is a little ambiguous, we support a PKIX DER encoded public key or a PEM encoded version of a PKIX key (Not for example a PKCS1 encoding, typically just used for RSA). I'd rather the API be "Just give us a public key, however you've generated it" and let the server handle different formats. It also allows us to support more formats without requiring an update to the API.

This was one reason I thought about supporting the CSR as a "public key" type, but I think being able to annotate all fields as required is more beneficial.

I'm adding a comment to the API to specify the supported key formats.

Reply:
Updated

Reply:
> I don't think this list will continue to grow.

Famous last words :)

What if we wanted to, say, issue SSH certs? [Smallstep](https://smallstep.com/blog/use-ssh-certificates) runs an SSO-based CA for SSH certificates, which I could imagine being something we want to bring to Fulcio (spoiler: I have some need for this). The SSH key formats (there are at least 2, wire and the stuff that goes in your `authorized_keys` file) look quite different from the above.

I suppose we could make another field in the `oneof` for that, or we could extend the "try every key format" if-ladder. I suppose both of those are compatible the code in this PR—if you agree then feel free to resolve this.

Reply:
SSH is an entirely different beast in my opinion, and I think it'd be best to tackle it as a separate gRPC service. I've in the past tried to add an SSH CA to an X.509 based CA API, and the design gets muddy quickly. For one, the PKI is different - no chains, just a root issuing a leaf. There's support for different public key types too.

Reply:
We'd also need support for SSH cert extensions to restrict how the cert is used, etc. If this was something you were interested in, let's chat more, I'd be happy to sketch out an API for SSH.